rs/ledger_suite/icrc1/ledger/src/lib.rs

```diff
@@ -883,3 +970,107 @@ impl<Tokens: TokensType> Ledger<Tokens> {
         }
     }
 }
+
+#[derive(Serialize, Deserialize, Debug, Default)]
+pub struct StableAllowancesData {}
+
+impl AllowancesData for StableAllowancesData {
+    type AccountId = Account;
+    type Tokens = Tokens;
+
+    fn get_allowance(
+        &self,
+        account_spender: &(Self::AccountId, Self::AccountId),
+    ) -> Option<Allowance<Self::Tokens>> {
+        ALLOWANCES_MEMORY.with_borrow(|allowances| allowances.get(account_spender))
+    }
+
+    fn set_allowance(
+        &mut self,
+        account_spender: (Self::AccountId, Self::AccountId),
+        allowance: Allowance<Self::Tokens>,
+    ) {
+        ALLOWANCES_MEMORY
+            .with_borrow_mut(|allowances| allowances.insert(account_spender, allowance));
+    }
+
+    fn remove_allowance(&mut self, account_spender: &(Self::AccountId, Self::AccountId)) {
+        ALLOWANCES_MEMORY.with_borrow_mut(|allowances| allowances.remove(account_spender));
+    }
+
+    fn insert_expiry(
+        &mut self,
+        timestamp: TimeStamp,
+        account_spender: (Self::AccountId, Self::AccountId),
+    ) {
+        ALLOWANCES_EXPIRATIONS_MEMORY.with_borrow_mut(|expirations| {
+            expirations.insert((timestamp, account_spender), ());
+        });
+    }
+
+    fn remove_expiry(
+        &mut self,
+        timestamp: TimeStamp,
+        account_spender: (Self::AccountId, Self::AccountId),
+    ) {
+        ALLOWANCES_EXPIRATIONS_MEMORY.with_borrow_mut(|expirations| {
+            expirations.remove(&(timestamp, account_spender));
+        });
+    }
+
+    fn insert_arrival(
+        &mut self,
+        _timestamp: TimeStamp,
+        _account_spender: (Self::AccountId, Self::AccountId),
+    ) {
+    }
+
+    fn remove_arrival(
+        &mut self,
+        _timestamp: TimeStamp,
+        _account_spender: (Self::AccountId, Self::AccountId),
+    ) {
+    }
+
+    fn first_expiry(&self) -> Option<(TimeStamp, (Self::AccountId, Self::AccountId))> {
+        ALLOWANCES_EXPIRATIONS_MEMORY
+            .with_borrow(|expirations| expirations.first_key_value().map(|kv| kv.0))
+    }
+
+    fn oldest_arrivals(&self, _n: usize) -> Vec<(Self::AccountId, Self::AccountId)> {
+        vec![]
```

Comment:
This value might not reflect previous function calls to `insert_arrival` and `remove_arrival` - can't this cause surprises?

Reply:
We do not store arrivals in stable structures, I added a comment. The arrivals were only used for trimming allowances. This function will be removed entirely when implementing V3 for the ICP ledger.

Reply:
> We do not store arrivals in stable structures, I added a comment.
Sure, makes sense.
> The arrivals were only used for trimming allowances. This function will be removed entirely when implementing V3 for the ICP ledger.
As long as that code is around, is there a possibility of a spurious failure due to returning fake data here?

Reply:
This function is only called by `select_approvals_to_trim` which in turn is only called by the trimming code https://github.com/dfinity/ic/blob/7927c348c7acd1bd6433342bf6ebc9f7b34e6f3f/rs/ledger_suite/common/ledger_canister_core/src/ledger.rs#L349 I left in the `test_approval_trimming` test (https://github.com/dfinity/ic/blob/7927c348c7acd1bd6433342bf6ebc9f7b34e6f3f/rs/ledger_suite/tests/sm-tests/src/lib.rs#L3336) to make sure this code is executed in tests. The difference now is that for the ICRC ledger it does not check if the approvals were actually trimmed (see below).